design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,410 @@
+  - rook # main command entry points for operators and daemons
+    - ceph
+    - minio
+  - rookflex
+- pkg
+  - apis
+    - rook.io # rook.io API group, additional groups would be sibling dirs
+      - v1alpha1 # existing version of alpha Rook API
+      - v1alpha2 # new version of alpha Rook API that includes new types
+  - client
+    - clientset # generated strongly typed client code to access Rook API
+  - daemon # daemons for each storage backend
+    - ceph
+    - minio
+  - operator # all orchestration logic
+    - cluster # custom controllers for each storage backend
+      - ceph
+      - minio
+```
+
+### Custom Types
+
+Custom resources in Kubernetes use the following naming and versioning convention:
+
+* Group: A collection of several related types that can be enabled/disabled as a unit in the API (e.g., `rook.io`)
+* Version: The API version of the group (e.g., `v1alpha1`)
+* Kind: The specific type within the API group (e.g., `cluster`)
+
+Putting this together with an example, the `cluster` kind from the `rook.io` API group with a version of `v1alpha1` would be referred to in full as `cluster.rook.io/v1alpha1`.
+
+### Versioning
+
+Versioning of custom resources defined by Rook is important, and we should carefully consider a design that allows resources to be versioned in a sensible way.
+Think of the following scenario where we have the Ceph and Minio backends both to a level of maturity that we have deemed `stable`.
+Then, if a new storage backend is incorporated into Rook with an early implementation, it should be able to start at `alpha` to properly indicate its nascent level of maturity and stability.
+Therefore, there is a need for each backend to be able to be versioned **independently** of other backends.
+
+Independent versioning is possible by making each storage backend a top level custom resource.
+The same can not be said for an approach of a single cluster CRD that has specific backends as sub-fields.
+To demonstrate this point, let's explore further the scenario above where we have a stable `cluster` CRD that supports `ceph` and `minio`.
+This stable CRD has dropped some properties and gained some properties during its maturity from `alpha` to `stable`.
+To add the support for the brand new storage backend `foostore`, it should be added to the `alpha` CRD because that properly represents its maturity.
+Therefore, we would have the following yaml:
+
+```yaml
+apiVersion: rook.io/v1
+kind: StorageCluster
+spec:
+  someStableField: "stableValue"
+  ceph:
+  mino:
+---
+apiVersion: rook.io/v1alpha1
+kind: StorageCluster
+spec:
+  someAlphaField: "alphaValue"
+  ceph:
+  minio:
+  foostore:
+```
+
+This results in an unfortunate situation for the new `foostore` backend.
+Because its version is coupled with that of the top level cluster CRD, and it is of `alpha` maturity, it is forced to be included in the old `alpha` cluster type, meaning it can't use the latest mature and stable fields.
```

Comment:
If `StorageCluster` is simultaneously (in a given Rook version) supported through both `rook.io/v1` and `rook.io/v1alpha1`, then k8s API mechanics require that both versions must have some place to store the `foostore` field. That means the opposite is actually true: you are forced to allow `foostore` to be used with `StorageCluster` in `rook.io/v1`. This is also bad (because you would rather limit it) but in a different way.

Reply:
This is also a hugely important clarification about the k8s API mechanics and has shaped the latest design.  Your expertise is very welcome.